parser/errors.go

```diff
@@ -0,0 +1,137 @@
+
+	ErrAmountMatchAmountMissing      = errors.New("amount is missing")
+	ErrAmountMatchAmountPopulated    = errors.New("amount is populated")
+	ErrAmountMatchUnexpectedSign     = errors.New("unexpected amount sign")
+	ErrAmountMatchUnexpectedCurrency = errors.New("unexpected currency")
+
+	ErrCoinActionMatchCoinChangeIsNil      = errors.New("coin change is nil")
+	ErrCoinActionMatchUnexpectedCoinAction = errors.New("unexpected coin action")
+
+	ErrEqualAmountsNoOperations = errors.New("cannot check equality of 0 operations")
+	ErrEqualAmountsNotEqual     = errors.New("amounts are not equal")
+
+	ErrOppositeAmountsSameSign       = errors.New("operations have the same sign")
+	ErrOppositeAmountsAbsValMismatch = errors.New("operation absolute values are not equal")
+
+	ErrEqualAddressesTooFewOperations = errors.New("cannot check equality of <= 1 operations")
+	ErrEqualAddressesAccountIsNil     = errors.New("account is nil")
+	ErrEqualAddressesAddrMismatch     = errors.New("addresses do not match")
+
+	ErrMatchIndexValidIndexOutOfRange = errors.New("match index out of range")
+	ErrMatchIndexValidIndexIsNil      = errors.New("match index is nil")
+
+	ErrMatchOperationsNoOperations          = errors.New("unable to match anything to 0 operations")
+	ErrMatchOperationsDescriptionsMissing   = errors.New("no descriptions to match")
+	ErrMatchOperationsMatchNotFound         = errors.New("unable to find match for operation")
+	ErrMatchOperationsDescriptionNotMatched = errors.New("could not find match for description")
+
+	MatchOpsErrs = []error{
+		ErrAccountMatchAccountMissing,
+		ErrAccountMatchSubAccountMissing,
+		ErrAccountMatchSubAccountPopulated,
+		ErrAccountMatchUnexpectedSubAccountAddr,
+		ErrMetadataMatchKeyNotFound,
+		ErrMetadataMatchValueTypeMismatch,
+		ErrAmountMatchAmountMissing,
+		ErrAmountMatchAmountPopulated,
+		ErrAmountMatchUnexpectedSign,
+		ErrAmountMatchUnexpectedCurrency,
+		ErrCoinActionMatchCoinChangeIsNil,
+		ErrCoinActionMatchUnexpectedCoinAction,
+		ErrEqualAmountsNoOperations,
+		ErrEqualAmountsNotEqual,
+		ErrOppositeAmountsSameSign,
+		ErrOppositeAmountsAbsValMismatch,
+		ErrEqualAddressesTooFewOperations,
+		ErrEqualAddressesAccountIsNil,
+		ErrEqualAddressesAddrMismatch,
+		ErrMatchIndexValidIndexOutOfRange,
+		ErrMatchIndexValidIndexIsNil,
+		ErrMatchOperationsNoOperations,
+		ErrMatchOperationsDescriptionsMissing,
+		ErrMatchOperationsMatchNotFound,
+		ErrMatchOperationsDescriptionNotMatched,
+	}
+)
+
+// ErrParser takes an error as an argument and returns
+// whether or not the error is one thrown by the asserter
+// along with the specific source of the error
+func ErrParser(err error) (bool, string) {
+	parserErrs := map[string][]error{
+		"intent error":           IntentErrs,
+		"match operations error": MatchOpsErrs,
+	}
+
+	for key, val := range parserErrs {
+		if findError(val, err) {
+			return true, key
+		}
+	}
+	return false, ""
+}
+
+// findError
+func findError(errorList []error, err error) bool {
```

Comment:
Can we make this function a util so you can use it across all the packages you are overhauling?